docs/design/sql/01-type-system.md

```diff
@@ -0,0 +1,157 @@
+# SQL Type System
+
+## Overview
+Hazelcast SQL type system defines how objects of different types interact with each other. Type system
+is defined by the list of supported types, type mapping and type conversion rules.
+
+Type is described by name, precedence, precision and scale.
+- Name is textual representation of type name
+- Precedence is an integer number which defines context-dependent type resolution rules for expressions
+- Precision is the total count of significant digits in the whole number, applicable to **numeric types**
+
+Type family is a collection of types with the same name, but different precisions. All types
+within a family have the same name and precedence. For example, `INT(11)` and `INT(12)` are two types
+from the same `INT` family.
+
+Scale is not used in Hazelcast SQL. It is applicable only for DECIMAL, REAL and DOUBLE types, and instead
+of defining it as a separate value, we just treat these types as types with infinite scale.
+
+## Supported types
+Types supported by the Hazelcast SQL engine is listed in the Table 1. Precision is the smallest precision in the type family.
+
+`OBJECT` is Hazelcast-specific type representing an object which doesn't match any other type.
+
+*Table 1: Hazelcast SQL Data Types*
+
+| SQL Type | Precedence | Precision |
+|---|---|---|
+| `VARCHAR` | 100 |  |
+| `BIT` | 200 | 1 |
+| `TINYINT` | 300 | 4 |
+| `SMALLINT` | 400 | 7 |
+| `INT` | 500 | 11 |
+| `BIGINT` | 600 | 20 |
+| `DECIMAL` | 700 | Unlimited |
+| `REAL` | 800 | Unlimited |
+| `DOUBLE` | 900 | Unlimited |
+| `TIME` | 1000 |  |
+| `DATE` | 1100 |  |
+| `TIMESTAMP` | 1200 |  |
+| `TIMESTAMP WITH TIMEZONE` | 1300 |  |
```

Comment:
Not a big deal, but in the standard it's named as `TIMESTAMP WITH TIME ZONE`, with a space between `TIME` and `ZONE`. Looks like other products are following that style as well.

Reply:
Renamed to `TIMESTAMP WITH TIME ZONE`